### PR TITLE
Enforce default discriminant type for all representations and across editions

### PR DESCRIPTION
**Pull request by daxpedda**
See [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/enum.20discriminant.20size/near/388771801).

Yanking time ... cry me a river ...

### CODE REVIEW

**Comment by ModProg** [comment]
we could do this for all cases right? even the non `C` one? I feel like that would make sense, as technically also the `repr(Rust)` could change on an edition.

**Comment by daxpedda** [comment]
I was having the same thought, but I would have said that would break all proc-macros out there, so we should be safe? I don't want to further ruin the code generation ...

WDYT?

**Comment by daxpedda** [comment]
We could also ask on Zulip again?

**Comment by ModProg** [comment]
well, I'd say even the change to `repr(C)` would break all proc macros.

File: src/test/basic.rs [hunk 1]
```diff
@@ -197,7 +197,7 @@ fn enum_() -> Result<()> {
 	};
 	#[cfg(not(feature = "nightly"))]
 	let ord = quote! {
-		fn __discriminant<T>(__this: &Test<T>) -> isize {
+		const fn __discriminant<T>(__this: &Test<T>) -> isize {
```
Comment (daxpedda) [inline src/test/basic.rs:200]: WDYT of that?

Comment (ModProg) [reply src/test/basic.rs:200]: Does a const function ensure what we want? i.e. make overflow a compilation error?

Comment (ModProg) [reply src/test/basic.rs:200]: I checked that it doesn't in #82

Comment (daxpedda) [reply src/test/basic.rs:200]: Split off to #82.

**Comment by daxpedda** [comment]
It's done, checking both now.

**Comment by daxpedda** [comment]
I think we can get this out first and figure out #83 later when we get an answer.

**Review by ModProg — APPROVED** [review]

**Pull request merged**